defi/amms/bonding/bonding_macros/src/lib.rs

```diff
@@ -0,0 +1,107 @@
+                    }
+                });
+                /*
+                let block = match method.sig.output {
+                    ReturnType::Default => { empty_block },
+                    ReturnType::Type(_, ref box_type) => {
+                        let is_decimal: bool = match **box_type {
+                            syn::Type::Path(ref type_path) => {
+                                type_path.path.is_ident(&decimal_ident)
+                            },
+                            _ => false
+                        };
+                        if is_decimal {
+                            TokenStream::from(quote! {
+                                { // block returning the default for Decimal
+                                    0.into()
+                                    //let result: #box_type = Default::default();
+                                    //result
+                                }
+                            })
+                        } else {
+                            TokenStream::from(quote! {
+                                { // block returning the default for the type
+                                    let result: #box_type = Default::default();
+                                    result
+                                }
+                            })
+                        }
+                    }
+                };
+                let block = parse_macro_input!(block as Block);
+                */
+                let block = parse_macro_input!(empty_block as Block);
+                method.default = Some(block.clone());
+                // now update the vec of ImplItem::Method from the TraitItem::Method 
+                the_impl.items.push(
+                    ImplItem::Method(ImplItemMethod {
+                        attrs: the_impl.attrs.clone(), // copy the #[allow(unused)] to each method
+                        vis: Visibility::Public(
+                            VisPublic{
+                                pub_token: Pub{
+                                    span: Span::call_site()
+                                }
+                            }
+                        ),
+                        defaultness: None,
+                        sig: method.sig.clone(),
+                        block: block
+                    }
+                )
+
+                );
+            },
+            _ => {}
+        }
+    }
+
+    // create the final output
+    // a blueprint! with the geenerated impl, but all inside a module so the code is not callable
+    // then reexport only the stubs
+    let mod_name = Ident::new(&format!("internal_{}", ident), Span::call_site());
+    TokenStream::from(quote!{
+        mod #mod_name {
+        use super::*;
+        blueprint! {
+            struct #ident {}
+            #the_impl
+        }
+        }
+        // reexport the stub only
+        pub use #mod_name::#ident;
+    })
+}
```

Comment:
This is an interesting idea - transforming a trait into a blueprint with empty implementation.  One suggestion if I can make is probably to use `#(#var)*` syntax from `quote` so you can avoid generating-and-parsing token stream.

Reply:
I really liked this concept of transforming the trait, but absolutely the implementation is just a first cut to have it work.  I took the approach of doing the least intrusive thing to reuse the most code, but I'm not sure the impact on the size of the generated code, if the release build really would be smart enough to remove all of the non-pub stuff during LTO, particularly because of the `extern "C"` it might not.  Also I didn't want to depend too much on the implementation of blueprint knowing it could change.

But if this kind of thing were part of Scrypto I would imagine that maybe a better way would be to duplicate some code from the existing `blueprint!` macro but only generate the minimal amount of API stubs, and not actually make the entire blueprint.  It could also generate a real Rust trait in such a way that a blueprint could implement it to get static checks that the "receiving" side of the function/method calls have all the right methods.

I'll open an enhancement issue to track this idea.